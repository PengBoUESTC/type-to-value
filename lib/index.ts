import {
  Project,
  type SourceFile,
  SyntaxKind,
  type Type,
  ts,
  type InterfaceDeclaration,
  type TypeAliasDeclaration,
  type EnumDeclaration,
} from 'ts-morph'
import { cwd } from 'node:process'
import { join } from 'node:path'

interface Options {
  /** cache TypeToValue instance or not */
  cache?: boolean
  /** file path need to parse */
  sourceFilePath: string
  /** ts confing file path */
  tsConfigFilePath?: string
}

interface ConvertConfig {
  [path: string]: any
}

let instanceCache: Record<string, TypeToValue> = {}

export const dropInsCache = (key?: string) => {
  if (!key) {
    instanceCache = {}
  } else {
    delete instanceCache[key]
  }
  if (process.env.BUILD) return
  return instanceCache
}

let convertCache: Record<string, any> = {}

export const dropConvertCache = (key?: string) => {
  if (!key) {
    convertCache = {}
  } else {
    delete convertCache[key]
  }
  if (process.env.BUILD) return
  return convertCache
}

export const genCacheKey = (options: Options): string => {
  const { sourceFilePath, tsConfigFilePath = '' } = options
  return `${sourceFilePath}_${tsConfigFilePath}`
}

export const genConvertCacheKey = (path: string, typeValue: string) => {
  return `${path}_${typeValue}`
}

const setCache = (options: Options, instance: TypeToValue): TypeToValue => {
  const key = genCacheKey(options)
  instanceCache[key] = instance
  return instance
}

const setConvertCache = (path: string, typeValue: string, value: any) => {
  convertCache[genConvertCacheKey(path, typeValue)] = value
  return value
}

const getCache = (options: Options) => {
  const cacheKey = genCacheKey(options)
  const instance = instanceCache[cacheKey]
  if (!instance) {
    return setCache(options, new TypeToValue({ ...options, cache: false }))
  }

  return instance
}

const getConvertCache = (path: string, typeValue: string) => {
  return convertCache[genConvertCacheKey(path, typeValue)]
}

export const createTypeToValue = (options: Options) => {
  const { cache } = options
  if (cache) {
    return getCache(options)
  }
  return new TypeToValue(options)
}

export class TypeToValue {
  private sourceFileCache: Record<string, SourceFile> = {}
  private typeKeyCount: Record<string, number> = {}
  project: Project | null = null
  constructor(options: Options) {
    if (options.cache) {
      return createTypeToValue(options)
    }
    this.sourceFileCache = {}
    this.typeKeyCount = {}
    this.project = this.init(options)
  }

  get keyCount() {
    return this.typeKeyCount
  }

  get sourceFilesPaths() {
    const result: Record<string, string> = {}
    Object.entries(this.sourceFileCache).map(([key, sourceFile]) => {
      result[key] = sourceFile.getFilePath()
    })

    return result
  }

  get sourceFiles() {
    return this.sourceFileCache
  }

  init(options: Options) {
    const { sourceFilePath, tsConfigFilePath = join(cwd(), 'tsconfig.json') } =
      options
    const project = new Project({
      tsConfigFilePath,
    })
    project.addSourceFilesAtPaths(sourceFilePath)
    const tsFiles = project
      .getDirectories()
      .map((d) => d.getSourceFiles().map((s) => s.getFilePath()))
      .filter((item) => item.length)
      .flat(1)

    tsFiles.forEach((tsFile) => {
      const sourceFiletest = project.getSourceFile(tsFile)
      if (!sourceFiletest) return
      const interfaceList = sourceFiletest.getInterfaces()
      const typeAliases = sourceFiletest.getTypeAliases()
      const enums = sourceFiletest.getEnums()
      const list = [...interfaceList, ...typeAliases, ...enums]
      if (list.length) {
        list.forEach((inter) => {
          const name = inter.getName()
          this.typeKeyCount[name] = (this.typeKeyCount[name] || 0) + 1

          this.sourceFileCache[`${name}-${this.typeKeyCount[name]}`] =
            inter.getSourceFile()
        })
      }
    })

    return project
  }

  generateValue(type: Type<ts.Type>, config?: ConvertConfig): any {
    if (type.isUndefined()) {
      return undefined
    }
    if (type.isNull()) {
      return null
    }

    if (type.isLiteral()) {
      return this.genLiteralValue(type)
    }

    if (type.isString()) {
      return 'string'
    }

    if (type.isBigInt()) {
      return BigInt('9007199254740991')
    }

    if (type.isNumber()) {
      return 0
    }
    if (type.isBoolean()) {
      return true
    }
    if (type.isEnum()) {
      return this.genEnum(
        type
          .getSymbol()
          ?.getDeclarations()[0]
          .asKindOrThrow(SyntaxKind.EnumDeclaration),
      )
    }
    if (type.isUnion()) {
      const unionTypes = type.getUnionTypes()
      return this.generateValue(
        unionTypes.find((t) => !t.isUndefined()) || unionTypes[0],
      )
    }
    if (type.isArray()) {
      const elementType = type.getArrayElementTypeOrThrow()
      return [this.generateValue(elementType)]
    }
    if (type.isTuple()) {
      const tupleElements = type.getTupleElements()
      return tupleElements.map((element) => this.generateValue(element))
    }
    if (type.isObject()) {
      return this.genInnerObject(type, config)
    }
    if (type.isVoid()) {
      return void 0
    }
    if (type.isUnknown()) {
      return {}
    }

    // handle symbol type
    if (type.getText() === 'symbol') {
      return Symbol('symbol')
    }
    const properties = type.getProperties()
    // 由外部导入的数据类型
    if (!properties.length) {
      return this.genOuterObject(type, config)
    }

    return null
  }

  genLiteralValue(type: Type<ts.Type>) {
    if (type.isBooleanLiteral()) {
      // 获取布尔字面量的值
      return type.getText() === 'true'
    }
    return type.getLiteralValue()
  }

  genEnum(enumDeclaration?: EnumDeclaration) {
    if (!enumDeclaration) return
    const members = enumDeclaration.getMembers()
    if (!members.length) return
    return members[0].getValue()
  }

  genInnerObject(type: Type<ts.Type>, config?: ConvertConfig) {
    const value: any = {}
    const properties = type.getProperties()
    properties.forEach((prop) => {
      const name = prop.getName()
      if (config && config.hasOwnProperty(name)) {
        value[name] = config[name]
      } else {
        const t = prop.getDeclarations()[0]
        if (!t) {
          return
        }
        const propType = prop.getTypeAtLocation(t)
        value[name] = this.generateValue(propType, this.getConfig(name, config))
      }
    })
    return value
  }

  genOuterObject(type: Type<ts.Type>, config?: ConvertConfig) {
    const name = type.getText()
    // 当 没有找到 正确的类型文件时，会全局找到 ts 文件中声明的类型作为 兜底补充
    // 正常场景不会走到 这种场景
    for (let i = 1; i <= this.typeKeyCount[name]; i++) {
      const sourceFile = this.sourceFileCache[`${name}-${i}`]
      if (sourceFile) {
        const interfaceDeclaration = (
          sourceFile.getInterface(name) ||
          sourceFile.getTypeAlias(name) ||
          sourceFile.getEnum(name)
        )?.getType()
        if (interfaceDeclaration) {
          return this.generateValue(interfaceDeclaration, config)
        }
      }
    }
    return {}
  }

  getConfig(leadKey: string, config?: ConvertConfig) {
    if (!config) return config
    const nextConfig: Record<string, any> = {}
    let isEmpty = true
    Object.keys(config).forEach((key) => {
      if (!key) return
      const keys = key.split('.')
      if (keys[0] !== leadKey) return
      const nextKey = keys.slice(1).join('.')
      if (!nextKey) return
      isEmpty = false
      nextConfig[nextKey] = config[key]
    })
    if (isEmpty) return undefined
    return nextConfig
  }

  runBase(
    interfaceDeclaration?: InterfaceDeclaration | TypeAliasDeclaration,
    config?: ConvertConfig,
  ) {
    if (!interfaceDeclaration) return {}
    return this.generateValue(interfaceDeclaration.getType(), config)
  }

  run(path: string, typeValue: string, config?: ConvertConfig) {
    const sourceFile = this.project?.getSourceFile(path)
    if (!sourceFile) return null
    return this.runBase(
      sourceFile.getInterface(typeValue) || sourceFile.getTypeAlias(typeValue),
      config,
    )
  }

  runWithCache(path: string, typeValue: string, config?: ConvertConfig) {
    const cacheResult = getConvertCache(path, typeValue)
    if (cacheResult) return cacheResult
    const result = this.run(path, typeValue, config)
    setConvertCache(path, typeValue, result)
    return result
  }

  runWithCopy(path: string, typeValue: string, config?: ConvertConfig) {
    const result = this.runWithCache(path, typeValue, config)

    return JSON.parse(JSON.stringify(result))
  }
}
