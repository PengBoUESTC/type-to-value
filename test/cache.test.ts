import { TypeToValue, createTypeToValue, genCacheKey, dropInsCache, genConvertCacheKey, dropConvertCache } from '../lib/index'

describe("cahce test", () => {
  test("createTypeToValue test", () => {
    const instance = createTypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    const instance2 = createTypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    const instance3 = createTypeToValue({ sourceFilePath: 'test/**/*.d.ts' })
    const instance4 = new TypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    expect(instance2 === instance).toBeTruthy()
    expect(instance3 === instance).toBeFalsy()
    expect(instance4 === instance).toBeTruthy()
  })

  test("dropInsCache test", () => {
    const options = { cache: true, sourceFilePath: 'test/**/*.d.ts' }
    const instance = createTypeToValue(options)
    dropInsCache(genCacheKey(options))
    const instance2 = createTypeToValue(options)
    expect(instance === instance2).toBeFalsy()
  })

  test("runWithCache test", () => {
    const instance = createTypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    const run = jest.spyOn(instance, 'run')

    const result = instance.run('test/dts/index.d.ts', 'Data')
    const result2 = instance.run('test/dts/index.d.ts', 'Data')
    expect(run).toHaveBeenCalledTimes(2)
    const resultCache = instance.runWithCache('test/dts/index.d.ts', 'Data')
    const resultCache2 = instance.runWithCache('test/dts/index.d.ts', 'Data')
    expect(run).toHaveBeenCalledTimes(3)
    expect(result === result2).toBeFalsy()
    expect(resultCache === resultCache2).toBeTruthy()
  })

  test("dropConvertCache test", () => {
    const instance = createTypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    const resultCache = instance.runWithCache('test/dts/index.d.ts', 'Data')
    dropConvertCache(genConvertCacheKey('test/dts/index.d.ts', 'Data'))
    const resultCache2 = instance.runWithCache('test/dts/index.d.ts', 'Data')
    expect(resultCache === resultCache2).toBeFalsy()
  })

  test("dropInsCache clean test", () => {
    const options = { cache: true, sourceFilePath: 'test/**/*.d.ts' }
    const instance = createTypeToValue(options)
    const instance2 = createTypeToValue({ ...options, sourceFilePath: 'test/**/array.d.ts' })
    const result = dropInsCache(genCacheKey(options))
    expect(result).toMatchSnapshot()
    const result2 = dropInsCache()
    expect(result2).toEqual({})
  })
  test("dropConvertCache clean test", () => {
    const instance = createTypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    const resultCache = instance.runWithCache('test/dts/index.d.ts', 'Data')
    const resultCache2 = instance.runWithCache('test/dts/index.d.ts', 'DataCache')
    
    const result = dropConvertCache(genConvertCacheKey('test/dts/index.d.ts', 'Data'))
    expect(result).toMatchSnapshot()
    const result2 = dropConvertCache()
    expect(result2).toEqual({})
  })

  test("runWithCopy test", () => {
    const instance = createTypeToValue({ cache: true, sourceFilePath: 'test/**/*.d.ts' })
    const resultCache = instance.runWithCache('test/dts/index.d.ts', 'Data')
    const resultCopy = instance.runWithCopy('test/dts/index.d.ts', 'Data')
    resultCache.a = -100
    expect(resultCopy).toMatchSnapshot()
    expect(resultCache).toMatchSnapshot()
    expect(resultCopy.a).not.toEqual(-100)
  })
})