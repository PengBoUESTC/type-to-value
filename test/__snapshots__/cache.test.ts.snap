// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cahce test dropConvertCache clean test 1`] = `
{
  "test/dts/index.d.ts_DataCache": {
    "a": true,
  },
}
`;

exports[`cahce test dropInsCache clean test 1`] = `
{
  "test/**/array.d.ts_": TypeToValue {
    "project": Project {
      "_context": ProjectContext {
        "compilerFactory": CompilerFactory {
          "documentRegistry": DocumentRegistry {},
        },
        "directoryCoordinator": DirectoryCoordinator {},
        "fileSystemWrapper": TransactionalFileSystem {},
        "inProjectCoordinator": InProjectCoordinator {},
        "lazyReferenceCoordinator": LazyReferenceCoordinator {},
        "logger": ConsoleLogger {},
        "manipulationSettings": ManipulationSettingsContainer {
          "_settings": {
            "indentationText": "    ",
            "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
            "newLineKind": 1,
            "quoteKind": """,
            "usePrefixAndSuffixTextForRename": false,
            "useTrailingCommas": false,
          },
        },
        "structurePrinterFactory": StructurePrinterFactory {
          "_getFormatCodeSettings": [Function],
        },
      },
    },
    "sourceFileCache": {
      "ArrayData-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 3,
      },
      "ArrayData2-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 3,
      },
      "ArrayData3-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 3,
      },
      "ConvertConfig-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 17,
      },
      "ImportMeta-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 2,
      },
      "ImportMetaEnv-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 2,
      },
      "Options-1": SourceFile {
        "__sourceFile": [Circular],
        "_context": ProjectContext {
          "compilerFactory": CompilerFactory {
            "documentRegistry": DocumentRegistry {},
          },
          "directoryCoordinator": DirectoryCoordinator {},
          "fileSystemWrapper": TransactionalFileSystem {},
          "inProjectCoordinator": InProjectCoordinator {},
          "lazyReferenceCoordinator": LazyReferenceCoordinator {},
          "logger": ConsoleLogger {},
          "manipulationSettings": ManipulationSettingsContainer {
            "_settings": {
              "indentationText": "    ",
              "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
              "newLineKind": 1,
              "quoteKind": """,
              "usePrefixAndSuffixTextForRename": false,
              "useTrailingCommas": false,
            },
          },
          "structurePrinterFactory": StructurePrinterFactory {
            "_getFormatCodeSettings": [Function],
          },
        },
        "_hasBom": undefined,
        "_inProject": true,
        "_referenceContainer": SourceFileReferenceContainer {},
        "_wrappedChildCount": 17,
      },
    },
    "typeKeyCount": {
      "ArrayData": 1,
      "ArrayData2": 1,
      "ArrayData3": 1,
      "ConvertConfig": 1,
      "ImportMeta": 1,
      "ImportMetaEnv": 1,
      "Options": 1,
    },
  },
}
`;

exports[`cahce test runWithCopy test 1`] = `
{
  "a": "string",
  "b": 0,
  "c": {},
}
`;

exports[`cahce test runWithCopy test 2`] = `
{
  "a": -100,
  "b": 0,
  "c": {},
}
`;
